test/unit/test_proxy.c

```diff
@@ -0,0 +1,430 @@
+#include "unittest.h"
+#include "../../src/net.h"
+#include "../../src/netutl.h"
+#include "../../src/proxy.h"
+#include "../../src/xalloc.h"
+
+static const char *user = "foo";
+static const size_t userlen = sizeof("foo") - 1;
+
+static const char *pass = "bar";
+static const size_t passlen = sizeof("bar") - 1;
+
+static int teardown(void **state) {
+	(void)state;
+
+	free(proxyuser);
+	proxyuser = NULL;
+
+	free(proxypass);
+	proxypass = NULL;
+
+	return 0;
+}
+
+static void test_socks_req_len_socks4_ipv4(void **state) {
+	(void)state;
+
+	const sockaddr_t sa = str2sockaddr("127.0.0.1", "4242");
+
+	size_t len = socks_req_len(PROXY_SOCKS4, &sa);
+	assert_int_equal(9, len);
+
+	proxyuser = xstrdup(user);
+	len = socks_req_len(PROXY_SOCKS4, &sa);
+	assert_int_equal(9 + userlen, len);
+}
+
+static void test_socks_req_len_socks4_ipv6(void **state) {
+	(void)state;
+
+	sockaddr_t sa = str2sockaddr("::1", "4242");
+	size_t len = socks_req_len(PROXY_SOCKS4, &sa);
+	assert_int_equal(0, len);
+}
+
+static void test_socks_req_len_socks5_ipv4(void **state) {
+	(void)state;
+
+	sockaddr_t sa = str2sockaddr("127.0.0.1", "4242");
+	size_t baselen = 13;
+
+	// Base test
+	size_t len = socks_req_len(PROXY_SOCKS5, &sa);
+	assert_int_equal(baselen, len);
+
+	// Setting only password must not change result
+	proxypass = xstrdup(pass);
+	len = socks_req_len(PROXY_SOCKS5, &sa);
+	assert_int_equal(baselen, len);
+
+	// Setting both must
+	proxyuser = xstrdup(user);
+	len = socks_req_len(PROXY_SOCKS5, &sa);
+	assert_int_equal(baselen + 3 + userlen + passlen, len);
+}
+
+static void test_socks_req_len_socks5_ipv6(void **state) {
+	(void)state;
+
+	sockaddr_t sa = str2sockaddr("::1", "4242");
+	size_t baselen = 25;
+
+	// Base test
+	size_t len = socks_req_len(PROXY_SOCKS5, &sa);
+	assert_int_equal(baselen, len);
+
+	// Setting only user must not change result
+	proxyuser = xstrdup(user);
+	len = socks_req_len(PROXY_SOCKS5, &sa);
+	assert_int_equal(baselen, len);
+
+	// Setting both must
+	proxypass = xstrdup(pass);
+	len = socks_req_len(PROXY_SOCKS5, &sa);
+	assert_int_equal(baselen + 3 + userlen + passlen, len);
+}
+
+static void test_socks_req_len_wrong_types(void **state) {
+	(void)state;
+
+	sockaddr_t sa = str2sockaddr("::1", "4242");
+
+	assert_int_equal(0, socks_req_len(PROXY_NONE, &sa));
+	assert_int_equal(0, socks_req_len(PROXY_SOCKS4A, &sa));
+	assert_int_equal(0, socks_req_len(PROXY_HTTP, &sa));
+	assert_int_equal(0, socks_req_len(PROXY_EXEC, &sa));
+}
+
+static void test_socks_req_len_wrong_family(void **state) {
+	(void)state;
+
+	sockaddr_t sa = {.sa.sa_family = AF_UNKNOWN};
+	assert_int_equal(0, socks_req_len(PROXY_SOCKS4, &sa));
+	assert_int_equal(0, socks_req_len(PROXY_SOCKS5, &sa));
+}
+
+static void test_check_socks_resp_wrong_types(void **state) {
+	(void)state;
+
+	uint8_t buf[512];
+	assert_false(check_socks_resp(PROXY_NONE, buf, sizeof(buf)));
```

Comment:
GCC 11.2.0 is correctly complaining that `buf[]` is used uninitialized here.

Reply:
Oops, somehow I missed that. Thanks.